dev/NumberBox/NumberBox.cpp

```diff
@@ -0,0 +1,387 @@
+#include "NumberBoxAutomationPeer.h"
+#include "RuntimeProfiler.h"
+#include "ResourceAccessor.h"
+#include "Utils.h"
+
+NumberBox::NumberBox()
+{
+    __RP_Marker_ClassById(RuntimeProfiler::ProfId_NumberBox);
+
+    // Default values for the number formatter
+    auto formatter = winrt::DecimalFormatter();
+    formatter.IntegerDigits(1);
+    formatter.FractionDigits(0);
+    NumberFormatter(formatter);
+
+    PointerWheelChanged({ this, &NumberBox::OnScroll });
+
+    SetDefaultStyleKey(this);
+}
+
+winrt::AutomationPeer NumberBox::OnCreateAutomationPeer()
+{
+    return winrt::make<NumberBoxAutomationPeer>(*this);
+}
+
+void NumberBox::OnApplyTemplate()
+{
+    winrt::IControlProtected controlProtected = *this;
+
+    if (auto spinDown = GetTemplateChildT<winrt::RepeatButton>(L"DownSpinButton", controlProtected))
+    {
+        spinDown.Click({ this, &NumberBox::OnSpinDownClick });
+
+        // Do localization for the down button
+        if (winrt::AutomationProperties::GetName(spinDown).empty())
+        {
+            auto spinDownName = ResourceAccessor::GetLocalizedStringResource(SR_NumberBoxDownSpinButtonName);
+            winrt::AutomationProperties::SetName(spinDown, spinDownName);
+        }
+    }
+
+    if (auto spinUp = GetTemplateChildT<winrt::RepeatButton>(L"UpSpinButton", controlProtected))
+    {
+        spinUp.Click({ this, &NumberBox::OnSpinUpClick });
+
+        // Do localization for the up button
+        if (winrt::AutomationProperties::GetName(spinUp).empty())
+        {
+            auto spinUpName = ResourceAccessor::GetLocalizedStringResource(SR_NumberBoxUpSpinButtonName);
+            winrt::AutomationProperties::SetName(spinUp, spinUpName);
+        }
+    }
+
+    m_textBox.set([this, controlProtected]() {
+        auto textBox = GetTemplateChildT<winrt::TextBox>(L"InputBox", controlProtected);
+        if (textBox)
+        {
+            textBox.LostFocus({ this, &NumberBox::OnTextBoxLostFocus });
+            textBox.KeyUp({ this, &NumberBox::OnNumberBoxKeyUp });
+        }
+        return textBox;
+    }());
+
+    // Initializing precision formatter. This formatter works neutrally to protect against floating point imprecision resulting from stepping/calc
+    m_stepPrecisionFormatter.FractionDigits(0);
+    m_stepPrecisionFormatter.IntegerDigits(1);
+    m_stepPrecisionFormatter.NumberRounder(nullptr);
+    m_stepPrecisionRounder.RoundingAlgorithm(winrt::RoundingAlgorithm::RoundHalfAwayFromZero);
+
+    SetSpinButtonVisualState();
+    UpdateTextToValue();
+}
+
+void NumberBox::OnValuePropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    auto oldValue = unbox_value<double>(args.OldValue());
+
+    if (std::isnan(Value()) && BasicValidationMode() == winrt::NumberBoxBasicValidationMode::InvalidInputOverwritten)
+    {
+        // In the validation case, we don't consider NaN to be valid.
+        Value(oldValue);
+    }
+
+    CoerseValue();
+
+    if (Value() != oldValue)
```

Comment:
>Value [](start = 8, length = 5)

This could have changed since you last checked for NAN.. do you need to check again?

Reply:
I moved things around a bit so this shouldn't be a problem anymore.